### ISSUE TITLE
[Question] Regarding the "Allow Updates", how to make it like the previous version?

### ISSUE THREAD

**Issue by cyfung1031** [violentmonkey/violentmonkey#1901]
Just a newbie to use Violentmonkey BETA. (v2.15.4)  

In previous version there is no such warning.

I want to edit the script downloaded from internet, but when there is any update, Violentmonkey can check for me and ask me should update it or not.

<img width="505" alt="Screen Shot 2023-09-22 at 10 31 34" src="https://github.com/violentmonkey/violentmonkey/assets/44498510/d1f47063-ca25-4651-b51a-2485b63a773d">

I think this behavior shall be still "allow update"? How to set it properly to make that allow me editing while notify me for the update where it is auto checked.

^^^^ this should be the default behavior as the same as Tampermonkey.

## Remarks

_Temporary Solution: https://github.com/violentmonkey/violentmonkey/issues/1901#issuecomment-1732253676_

**Comment by tophf** [comment]
No, it's entirely unreasonable to edit a script and let it auto-update. Either don't edit it or don't update.

**Comment by tophf** [comment]
We may change this behavior only when we implement #1023.

**Comment by cyfung1031** [comment]
> No, it's entirely unreasonable to edit a script and let it auto-update. Either don't edit it or don't update.

I am not asking for auto update but it should notify where there is an update, and prompt the message to the user where you want to update to the official version or keep your own version. 
If user decides to keep his own version, then no more update in the future will be asked. (even this can be decided by the user during the time he refused to update to the latest version)

For example, there is a userscript that it is useful but buggy, then the user might find some way to modify by himself.
after a year, the author come back and update the script to fix this bug, the user shall be able to know this, not just keep his own version forever.

**Comment by tophf** [comment]
You didn't understand. We don't have a separate check for updates. It'll be implemented in #1023. See also the workaround I added to my comment.

**Comment by F9y4ng** [comment]
After creating a new script locally why can't I edit it again, I have to manually close the update before I can edit it?

This is a terrible feature set.

**Comment by cyfung1031** [comment]
> After creating a new script locally why can't I edit it again, I have to manually close the update before I can edit it?
> 
> This is a terrible feature set.

@tophf  
> If you really want to shoot yourself in the foot, just uncheck the option, save, make your edits, save, enable the option again.

This is incorrect.
The correct procedure is uncheck the option, save, **REFRESH**, make your edits, save, enable the option again.

The saved option check would not apply immediately. It requires REFRESH.

@F9y4ng  but in another way, you can uncheck it, refresh, check it, save
Then the file can become editable + updatable HAHA
just don't refresh the page you can edit it with the option is checked.

**Comment by F9y4ng** [comment]
> @F9y4ng but in another way, you can uncheck it, save, refresh, check it, save Then the file can become editable + updatable HAHA just don't refresh the page you can edit it with the option is checked.

Why should I do these meaningless and redundant operations? 😂

This feature setting is so unfriendly to script author, and it becomes more difficult for normal users to modify the script temporarily.

**Comment by cyfung1031** [comment]
If you use Firefox, you can add the following as a bookmark.
```js
javascript:((a,b,c)=>{ c=a.checked; Promise.resolve().then(()=>a.click()).then(()=>b.click()).then(()=>setTimeout(()=>c&&location.reload(),100))  })(document.querySelector(".edit-settings.edit-body")?.querySelector("label")?.querySelector(`input[type="checkbox"]`),document.querySelector(".edit-buttons > button:first-child"));
```
Click Twice when you want to edit a script :)

**Comment by tophf** [comment]
> After creating a new script locally why can't I edit it again, I have to manually close the update before I can edit it?

Please describe the sequence of actions in more detail instead of emotionally discharging. It's not the end of the world and we can certainly find a better solution.

> This is a terrible feature set.

Just because something didn't work for you doesn't mean it's a bad feature. By design it's a good feature that protects the majority of users from shooting themselves in the leg unknowingly. The implementation may be bugged, of course, which we can fix and improve. This is why it's important to be very precise in your description of the steps necessary to reproduce the problem.

**Comment by F9y4ng** [comment]
- I already mentioned two options in #1902, which I replied to the wrong ISSUE:

  1. Keeping the habit of using it before v2.15.3, manually editing the local code will not affect the auto-update, and the auto-update overwrites the locally edited code.
  2. Adopting an approach similar to Tampermonkey 4.19, automatic updates are automatically lost when manually editing and saving local code, and manual operation is required to restore automatic updates.
  BTW. In the latest version of Tampermonkey4.20.BETA, it also uses **option 1** and gives a prompt that the local code has been edited.

- When creating a new script locally, for example:
```js
// ==UserScript==
// @name        test
// @namespace   test
// @match       *://*/*
// @version     1.0
// @author      You
// @description test
// @updateURL   https://example.com/test.meta.js
// @downloadURL https://example.com/test.user.js
// @grant       none
// ==/UserScript==

(function(){
  // do something...
})()
```
Then save&close it, after going to the website test script feature, when reopen the editor you will not be able to edit the code, you have to manually turn off the update before you can edit it. When I first saw that prompt, I was completely overwhelmed, it was so weird.

- As @cyfung1031 said, the current functionality is completely out of line with the user's operating habits when the user needs to temporarily modify the code to fix an urgent bug and wants to automatically update to the new version in a timely manner after the author updates the code.

- Objectively, this is not **A New Feature**,  but a sudden change to the editing rules that have been in use since v2.3.0. It is a very bad decision to take the risk of changing the user's usual habits. And if you still think it is right, OK, go ahead and do it. I will retract all my proposals above.

**Comment by cyfung1031** [comment]
@F9y4ng I think everything is already discussed in #1023.

### Summary
1. The concept of Auto Update is buggy and addressed since 2020.
2. Now violentmonkey made it read-only until the solution is done.
3. violentmonkey needs to design new UI and mechanism to make notification.
4. that's why the feature is planned but not yet done.

**Comment by F9y4ng** [comment]
> @F9y4ng I think everything is already discussed in #1023.
> 
> ### Summary
> 1. The current concept of Auto Update is glitchy and addressed since 2020.
> 2. Now violentmonkey made it read-only in 2.15.4 until the solution is done. (timeline TBC)
> 3. violentmonkey needs to design new UI and mechanism to make notification.
> 4. that's why the feature is planned but not yet done.

Thanks for your reply and summary.

My point is simple, don't take the risk of changing the user's habits, even if it might have some problems. because it's costly to do so.

**Comment by cyfung1031** [comment]
> > @F9y4ng I think everything is already discussed in #1023.
> > ### Summary
> > 
> > 1. The current concept of Auto Update is glitchy and addressed since 2020.
> > 2. Now violentmonkey made it read-only in 2.15.4 until the solution is done. (timeline TBC)
> > 3. violentmonkey needs to design new UI and mechanism to make notification.
> > 4. that's why the feature is planned but not yet done.
> 
> Thanks for your reply and summary.
> 
> My point is simple, don't take the risk of changing the user's habits, even if it might have some problems. because it's costly to do so.

See my temporary solution above.

**Comment by tophf** [comment]
Well, it's obviously a bug, which I didn't notice before, so naturally I will fix it now. You really don't have to add weight and ominous to your reports.

**Comment by tophf** [comment]
...and I'll auto-disable `[ ] Update` option for the scripts created locally, which would solve the bug for authors who create their publicly hosted scripts inside Violentmonkey and not in a file.

**Comment by tophf** [comment]
...probably not, because of the majority of users who are not the authors but simply copypasted the code with `@updateURL` from another site into Violentmonkey editor and saved it. The update should be enabled in such cases, and this is the majority of the users.

**Comment by cyfung1031** [comment]
for ui,
add a icon in each row


<img width="323" alt="Screen Shot 2023-09-23 at 22 17 25" src="https://github.com/violentmonkey/violentmonkey/assets/44498510/214dc0f4-ec75-4909-a4d1-83443ee20371">


status 1 - local script
status 2 - network script
status 2u - network script + update available
status 3 - network script + modified
status 3u - network script + modified + update available


before graphic designer comes, let assume these five icons as our base for discussion.

https://chat.openai.com/share/49013233-59c7-4816-b49b-2cb17a69a356


<img src="https://github.com/violentmonkey/violentmonkey/assets/44498510/7fe10fc3-e749-45de-a8c8-246944c408ca" width="16">
<img src="https://github.com/violentmonkey/violentmonkey/assets/44498510/734e3e2f-c136-4b1f-94e8-1839b1ba81ba" width="16">
<img src="https://github.com/violentmonkey/violentmonkey/assets/44498510/9de5f180-b068-44ea-86d5-5d906633b352" width="16">

<img src="https://github.com/violentmonkey/violentmonkey/assets/44498510/0fd69844-3d42-4830-b0a7-89bb9db74ff6" width="16">
<img src="https://github.com/violentmonkey/violentmonkey/assets/44498510/c85dfbe4-a0e5-462c-b40a-c22167f9099b" width="16">


When the upgrade arrow is shown, the user can click on it (hover is pointer) to update the script.

there is also option to auto upgrade. user can set the case
default is network script + update available will do auto update.
other case can be , for example, modified file - throw the old script to rubbish bin and update.

do the same for setting page. just enlarge the icon.

when update available is detected, change the number in the toolbar icon to up arrow.
it will show what it should show PLUS new update available (might or might not applied in the current webpage)
after clicked, and closed, the up arrow will be removed and revert to the number (indicating number of scripts activated in the current page)

**Comment by tophf** [comment]
Our goal is a lightweight extension. Violentmonkey was created in part to get rid of the gaudy overloaded interface in Tampermonkey. These icons aren't necessary for the majority of users who don't edit remotely updated scripts and who are just fine with installing a script and letting it auto-update.

**Comment by cyfung1031** [comment]
> Our goal is a lightweight extension. Violentmonkey was created in part to get rid of the gaudy overloaded interface in Tampermonkey. These icons aren't necessary for the majority of users who don't edit remotely updated scripts and who are just fine with installing a script and letting it auto-update.

users can know which script have update. if all scripts are just downloaded from internet. yes, all scripts will have the same symbol.
you cannot avoid to implement notification for update check.
Or you do the way same as Tampermonkey to popup whenever update is available which is annoying.

**Comment by tophf** [comment]
As I already noted, we don't really have to change our current auto-update behavior as it's arguably good for 99.9% of users judging by the fact that Chrome automatically updates the extensions without prompting first and extensions are used by hundreds of millions of users over the past decade. We'll probably implement it eventually, but it's not a priority.

**Comment by cyfung1031** [comment]
> As I already noted, we don't really have to change our current auto-update behavior as it's arguably good for 99.9% of users judging by the fact that Chrome automatically updates the extensions without prompting first and extensions are used by hundreds of millions of users over the past decade. We'll probably implement it eventually, but it's not a priority.

If the script is downloaded and no modified, default action can be still auto update.

**Comment by tophf** [comment]
You're just repeating the same argument over and over again, which I understood the first time. As I said it's an edge case. We'll try to improve it but it's not a priorty.

**Comment by tophf** [comment]
Here's the new UI:

![image](https://github.com/violentmonkey/violentmonkey/assets/1310400/18c9051d-8deb-4a02-a32b-fd510c48bee7)

**Comment by cyfung1031** [comment]
> Here's the new UI:
> 
> ![image](https://user-images.githubusercontent.com/1310400/270139874-ec0786f2-54b2-43ec-9a9e-9409306e8016.png)
> 
> ![image](https://user-images.githubusercontent.com/1310400/270139891-8034ff6a-487f-4c4a-aa5b-0e9fb6aa06f1.png)

Any option in Settings can tick "Allow edits" by default for any new downloaded script?

**Comment by tophf** [comment]
Currently it's not planned because it's still unclear to me why anyone would want to edit auto-updatable scripts.

**Comment by F9y4ng** [comment]
I updated to v2.15.5 BETA today, the experience is much better than before.

In the code setting page, **use global settings** cannot be selected when saving. Is this intentional or a bug?

**Comment by tophf** [comment]
Please provide more info because it certainly works for me.

**Comment by F9y4ng** [comment]
> Please provide more info because it certainly works for me.

![cam](https://github.com/violentmonkey/violentmonkey/assets/15855068/dd366290-faae-4796-870e-0ddfcebc9bcb)

**Comment by tophf** [comment]
Well, it works for me and I don't know how to reproduce the problem... Can you try in a new test profile?

**Comment by tophf** [comment]
Maybe the previous beta has corrupted the database.
Run the following in devtools console of the editor and verify the value of `config.notifyUpdates` is a number 0 or 1 or null:
```js
await browser.storage.local.get('scr:'+location.hash.match(/\d+/)[0])
```

**Comment by tophf** [comment]
If this is indeed the case, you can fix it by running the following in the editor, then reload the tab:
```js
await browser.runtime.sendMessage({
  cmd: 'UpdateScriptInfo',
  data: {
    config: {notifyUpdates: null},
    id: +location.hash.match(/\d+/)[0],
  },
})
```
The next release will be resistant to this problem.

**Comment by F9y4ng** [comment]
I deleted the 2.15.5BETA installed from the 2.15.4BETA upgrade, and reinstalled it (2.15.5BETA) from the chrome webstore, but the problem can still be reproduced.

**scr:** does not have the value `config.notifyUpdates`
```JSON
{
    "scr:1": {
        "config": {
            "enabled": 1,
            "removed": 0,
            "shouldUpdate": 2
        }
    }
}
```
`notifyUpdates: null` is successfully inserted when running `browser.runtime.sendMessage`, but **use global settings** is still not selected after refreshing the tab, but the value `notifyUpdates: null` exists at this time. 

When manually select **use global settings** again and save, the selection will still disappear and the value `notifyUpdates: null` will lost.

Looking forward to the next update will fix it.

**Comment by tophf** [comment]
The next update won't fix this, because lack of value is already covered by `==null`, which means the problem in your case is still unknown and I don't know what causes it.

**Comment by F9y4ng** [comment]
`on`, `off` are correct, but `use global settings` does not save the data correctly, nor does it read the data correctly, even though the values `notifyUpdates: null`​ ​have been inserted correctly.

Hope this information may help you.

**Comment by tophf** [comment]
No, it doesn't, because it works for me and I don't see in the source code how this problem can even exist.

**Comment by tophf** [comment]
DAMN! I've fixed it in c187f8e2 but I forgot that it wasn't included in 2.15.5, sorry for the confusion. The next release will be definitely fixed.

**Comment by F9y4ng** [comment]
> DAMN! I've fixed it in c187f8e2 but I forgot that it wasn't included in 2.15.5, sorry for the confusion. The next release will be definitely fixed.

All right, that's good, thanks.

**Comment by cyfung1031** [comment]
> DAMN! I've fixed it in [c187f8e](https://github.com/violentmonkey/violentmonkey/commit/c187f8e214a016a8367496831571bb1e52e475d5) but I forgot that it wasn't included in 2.15.5, sorry for the confusion. The next release will be definitely fixed.

Please build BETA 2.15.6 coz those 17 commits after 2.15.5 can really fix the bugs.

**Comment by tophf** [comment]
I only release quickly when there are critical bugs. You can use CI builds until then.
